.github/workflows/rust.yml

```
@@ -6,32 +6,98 @@ on:
   pull_request:
     branches: [ master ]
 
+concurrency:
+  group: rust-${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
 jobs:
   lint:
     runs-on: ubuntu-latest
+    timeout-minutes: 20
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          components: clippy,rustfmt
       - uses: Swatinem/rust-cache@v2
+
       - name: Build
-        run: cargo build --features protobuf-src
-      - name: Check Clippy
-        run: |
-          cargo clippy --tests --features telemetry,protobuf-src -- -D warnings
-          cargo clippy --tests --no-default-features --features compression,tokio-rustls-runtime,async-std-rustls-runtime,auth-oauth2,telemetry,protobuf-src -- -D warnings
-      - name: Install nightly rustfmt
-        run: rustup toolchain install nightly --component rustfmt
+        run: cargo build --features protobuf-src --locked
```

Adding `--locked` enforces using Cargo.lock for reproducible dependency resolution and fails fast if the lockfile is out of date.